File: src/SSHServer/MACAlgorithms/HMACSHA1.cs
using SSHServer.Packets;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace SSHServer.MACAlgorithms
{
    public class HMACSHA1 : IMACAlgorithm
    {
        System.Security.Cryptography.HMACSHA1 m_HMAC = null;

        public uint DigestLength
        {
            get
            {
                // https://tools.ietf.org/html/rfc4253#section-6.4
                // According to this, the DigestLength is 20
                return 20;
            }
        }

        public uint KeySize
        {
            get
            {
                // https://tools.ietf.org/html/rfc4253#section-6.4
                // According to this, the KeySize is 20
                return 20;
            }
        }

        public string Name
        {
            get
            {
                return "hmac-sha1";
            }
        }

        public byte[] ComputeHash(uint packetNumber, byte[] data)
        {
            if (m_HMAC == null)
                throw new SSHServerException(DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED, "SetKey must be called before attempting to ComputeHash.");

            using (ByteWriter writer = new ByteWriter())
            {
                writer.WriteUInt32(packetNumber);
                writer.WriteRawBytes(data);
                return m_HMAC.ComputeHash(writer.ToByteArray());
            }
        }

        public void SetKey(byte[] key)
        {
            m_HMAC = new System.Security.Cryptography.HMACSHA1(key);
        }
    }
}

File: SSHServerNodeJS/SSHServerNodeJS/MACAlgorithms/HMACSHA1.ts
import { IMACAlgorithm } from "./IMACAlgorithm";
import * as Exceptions from "../SSHServerException";
import { ByteWriter } from "../ByteWriter";

import crypto = require("crypto");

export class HMACSHA1 implements IMACAlgorithm {
    private m_MHAC: crypto.Hmac;
    private m_Key: Buffer;

    constructor() {
    }

    public getName(): string {
        return "hmac-sha1";
    }

    public getKeySize(): number {
        // https://tools.ietf.org/html/rfc4253#section-6.4
        // according to this, the KeySize is 20
        return 20;
    }

    public getDigestLength(): number {
        // https://tools.ietf.org/html/rfc4253#section-6.4
        // according to this, the DigestLength is 20
        return 20;
    }

    public setKey(key: Buffer): void {
        this.m_Key = key;
    }

    public computeHash(packetNumber: number, data: Buffer): Buffer {
        if (this.m_Key === null) {
            throw new Exceptions.SSHServerException(
                Exceptions.DisconnectReason.SSH_DISCONNECT_KEY_EXCHANGE_FAILED,
                "SetKey must be called before attempting to ComputeHash.");
        }

        let writer: ByteWriter = new ByteWriter();
        writer.writeUInt32(packetNumber);
        writer.writeRawBytes(data);

        let hmac: crypto.Hmac = crypto.createHmac("sha1", this.m_Key);
        hmac.update(writer.toBuffer());
        return hmac.digest();
    }
}
